Failures in `test_restart_endpoint_after_switch_wal`

This test is failing ~1.5% of the time.

It was added in https://github.com/neondatabase/neon/pull/8943

Examples:

## Case 1

https://neon-github-public-dev.s3.amazonaws.com/reports/pr-6560/11150102484/index.html#testresult/32593da5b1e125d1/retries

```
RuntimeError: Run ['/tmp/neon/bin/neon_local', 'endpoint', 'start', '--safekeepers', '1', 'ep-2'] failed:
  stdout:
    Starting existing endpoint ep-2...
    Starting postgres node at 'postgresql://cloud_admin@127.0.0.1:31674/postgres'
    SIGKILL & wait the started process
  stderr:
    command failed: timed out waiting to connect to compute_ctl HTTP

    Caused by:
        0: error sending request for url (http://127.0.0.1:31675/status)
        1: client error (Connect)
        2: tcp connect error: Connection refused (os error 111)
        3: Connection refused (os error 111)
```

## Case 2

https://neon-github-public-dev.s3.amazonaws.com/reports/pr-9232/11153678585/index.html#/testresult/ace582de47d6f7a9

```
test_runner/regress/test_wal_acceptor.py:1076: in test_restart_endpoint_after_switch_wal
    endpoint.safe_psql("SELECT 'works'")
test_runner/fixtures/neon_fixtures.py:330: in safe_psql
    return self.safe_psql_many([query], **kwargs)[0]
test_runner/fixtures/neon_fixtures.py:340: in safe_psql_many
    with closing(self.connect(**kwargs)) as conn:
test_runner/fixtures/neon_fixtures.py:284: in connect
    conn: PgConnection = psycopg2.connect(**self.conn_options(**kwargs))
/github/home/.cache/pypoetry/virtualenvs/non-package-mode-_pxWMzVK-py3.9/lib/python3.9/site-packages/psycopg2/__init__.py:122: in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
E   psycopg2.OperationalError: connection to server at "localhost" (127.0.0.1), port 28828 failed: server closed the connection unexpectedly
E   	This probably means the server terminated abnormally
E   	before or while processing the request.
```

--- a/safekeeper/src/receive_wal.rs
+++ b/safekeeper/src/receive_wal.rs
@@ -498,21 +498,18 @@ impl WalAcceptor {
         // we will send keepalives by replying to these requests once per second.
         let mut next_keepalive = Instant::now();
 
-        loop {
-            let opt_msg = self.msg_rx.recv().await;
-            if opt_msg.is_none() {
-                return Ok(()); // chan closed, streaming terminated
-            }
-            let mut next_msg = opt_msg.unwrap();
-
+        while let Some(mut next_msg) = self.msg_rx.recv().await {
             // Update walreceiver state in shmem for reporting.
             if let ProposerAcceptorMessage::Elected(_) = &next_msg {
                 walreceiver_guard.get().status = WalReceiverStatus::Streaming;
             }
 
             let reply_msg = if matches!(next_msg, ProposerAcceptorMessage::AppendRequest(_)) {
-                // loop through AppendRequest's while it's readily available to
-                // write as many WAL as possible without fsyncing
+                // Loop through AppendRequests while available to write as many WAL records as
+                // possible without fsyncing.
+                //
+                // Make sure the WAL is flushed before returning, see:
+                // https://github.com/neondatabase/neon/issues/9259
                 //
                 // Note: this will need to be rewritten if we want to read non-AppendRequest messages here.
                 // Otherwise, we might end up in a situation where we read a message, but don't
@@ -522,7 +519,7 @@ impl WalAcceptor {
 
                     if let Some(reply) = self.tli.process_msg(&noflush_msg).await? {
                         if self.reply_tx.send(reply).await.is_err() {
-                            return Ok(()); // chan closed, streaming terminated
+                            break; // disconnected, flush WAL and return on next send/recv
                         }
                     }
 
@@ -531,11 +528,13 @@ impl WalAcceptor {
                         break;
                     }
 
+                    // continue pulling AppendRequests if available
                     match self.msg_rx.try_recv() {
                         Ok(msg) => next_msg = msg,
                         Err(TryRecvError::Empty) => break,
-                        Err(TryRecvError::Disconnected) => return Ok(()), // chan closed, streaming terminated
-                    }
+                        // on disconnect, flush WAL and return on next send/recv
+                        Err(TryRecvError::Disconnected) => break,
+                    };
                 }
 
                 // flush all written WAL to the disk
@@ -555,5 +554,6 @@ impl WalAcceptor {
                 next_keepalive = Instant::now() + KEEPALIVE_INTERVAL;
             }
         }
+        Ok(())
     }
 }

